### ISSUE TITLE
Incompatible with Deno 1.4.[0,1,2] due to deno adopting tsconfig settings importsNotUsedAsValues and  isolatedModules 

### ISSUE THREAD

**Issue by luke-john** [yargs/yargs#1771]
In version 1.4 Deno adopted;

- ~the tsconfig setting [importsNotUsedAsValues](https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues) - https://github.com/denoland/deno/pull/7413~
  This seems to have been reverted in deno 1.4.5, and it appears they aren't planning on making it a default anymore https://github.com/denoland/deno/pull/7800
- the tsconfig setting [isolatedModules](https://www.typescriptlang.org/tsconfig#isolatedModules) when using the deno `--unstable` flag - https://github.com/denoland/deno/issues/7326 ([intended to be enabled by default in 1.5](https://github.com/denoland/deno/issues/7326))

These require the use of;
- type only imports for values which are only used as types (`importsNotUsedAsValues`)
- type only exports for types which are re-exported (`isolatedModules`)

From the errors I'm seeing with our tool which makes use of yargs from deno, this will require updates to the types for yargs, y18n and yargs_parser along the following lines.

```diff
# https://deno.land/x/y18n@v5.0.0-deno/deno.ts:2:1
- import { Y18NOpts } from './build/lib/index.d.ts'
+ import type { Y18NOpts } from './build/lib/index.d.ts'
```

```diff
# https://deno.land/x/yargs@v16.0.3-deno/types.ts#L6
// expose types for the benefit of Deno.
- import { YargsInstance as YargsType, Arguments } from './build/lib/yargs-factory.d.ts'
+ import type { YargsInstance as YargsType, Arguments } from './build/lib/yargs-factory.d.ts'

- export {
+export type {
  Arguments,
  YargsType
}
```

While there is a possible workaround of overriding these in a custom tsconfig, from 1.5 deno have indicated they will ignore any custom value for importsNotUsedAsValues from 1.5 onwards https://github.com/denoland/deno/pull/7413#issuecomment-691127947.

I'll open Pull Requests for these shortly, just getting the repos setup locally.

**Comment by bcoe** [comment]
@luke-john thank you for the patches 👌 

Do you think it would be worth updating our testing matrix too, to move to Deno 1.4, so that we'd catch this?

**Comment by bcoe** [comment]
@luke-john unrelated, there are a few shortcomings in Deno I'd love to address over time:

1. I don't think I picked an ideal default `bin` name, since it seems like folks install Deno bins in a variety of ways (_so `Deno run` isn't always how a command is run_).
2. I need to figure out a new approach to `requireDirectory`, that works with ESM -- I was thinking this might come in the next major, and involve yargs moving towards `async`/`await` for its parse.
3. We don't currently detect console width when doing layout, I believe Deno had plans to add a method for this.

Long story short, if you're actively using this in a tool you're building, would love help patching any shortcomings you're seeing.

**Comment by luke-john** [comment]
> Do you think it would be worth updating our testing matrix too, to move to Deno 1.4, so that we'd catch this?

I may be misunderstanding what you're referring to here, but from the looks of things the github action for deno is setup to use 1.x. so by default it picks the latest (https://github.com/yargs/yargs/pull/1772/checks?check_run_id=1188380822)

I've just done some testing, and it appears `deno test` behaves different to `deno run` and `deno test` is not impacted by the code which causes deno run to fail. So those tests pass with 1.4 currently.

It's end of day for me, but I'll have a look into this tomorrow and see if it's expected from `deno`s end, and if not open an issue there.

(also looking into this has made me realise I misunderstood how the deno branch is setup, and I should have branched the branches in the prs from master)

> 1. I don't think I picked an ideal default bin name, since it seems like folks install Deno bins in a variety of ways 

I don't have much of an opinion on this. The tool we're using it with is bundled up using [warp](https://github.com/dgiagio/warp), as it's consumed in environments where deno does not exist, so I'd manually set `scriptName`.   It's also our only "production" deno tool, I've otherwise only been using deno for one off scripts and tools.

> 2. I need to figure out a new approach to requireDirectory, that works with ESM -- I was thinking this might come in the next major, and involve yargs moving towards async/await for its parse.

I've not used the [commandDir](https://github.com/yargs/yargs/blob/master/docs/advanced.md#commanddirdirectory-opts) feature of yargs before, so I'm very familiar with how it's used.  I suspect it's will be less preferred by deno users, as it means you opt out of the possibility of inferring the types derived from yargs.  While this isn't quite there currently, it seems like the direction it makes sense to take the types towards (not a criticism, yargs is great as it is).

Saying that, an async/await strategy seems very workable with deno.

> 3. We don't currently detect console width when doing layout, I believe Deno had plans to add a method for this.

Just had a look into this, the api for this seems to still be be in the unstable lib https://github.com/denoland/deno/blob/master/cli/dts/lib.deno.unstable.d.ts#L46-L59

> 4. I wasn't sure how best to deal with Deno.env and fs-access errors -- yargs has features that use these features, but I didn't want to fail running a simple command line app if you're avoiding these sandboxed features.

Our binary required both fs and env permissions pre use of yargs, so didn't notice how this worked while getting started.

There appears to be an unstable api for querying/requesting Deno permissions https://github.com/denoland/deno/blob/master/cli/dts/lib.deno.unstable.d.ts#L1094-L1105.

This looks like it would solve this in a nice way, albeit it's also async so perhaps best considered for v2.

**Comment by bcoe** [comment]
@luke-john whoops, slight issue with:

https://github.com/yargs/yargs-parser/pull/329

Mind recreating the PR with the PR being against `master`. The `deno` branch is automatically generated whenever we perform a release.

**Comment by luke-john** [comment]
> I've just done some testing, and it appears deno test behaves different to deno run and deno test is not impacted by the code which causes deno run to fail. So those tests pass with 1.4 currently.

Did some more checking here, and it seems that the issues I've reported here actually only present when using 1.4.x with the unstable flag (currently intended to be present without from 1.5).

So updating the ci to run `deno --unstable test` would result in these issues being detected, but `deno test` will pass.

Having the tests run with the `--unstable` flag seems a little unusual, not sure if you want me to make that change.

**Comment by luke-john** [comment]
Resolved in https://github.com/yargs/yargs/pull/1753

Cheers for you help on this 👍
